vm/src/warn.rs

```diff
@@ -31,19 +62,130 @@ pub fn warn(
     )
 }
 
+fn get_filter(
+    _category: PyObjectRef,
+    _text: PyObjectRef,
+    _lineno: usize,
+    _module: PyObjectRef,
+    mut _item: PyObjectRef,
+    vm: &VirtualMachine,
+) -> PyResult<()> {
+    let filters = vm.state.warnings.filters.as_object().to_owned();
+
+    let filters: PyListRef = filters
+        .try_into_value(vm)
+        .map_err(|_| vm.new_value_error("_warnings.filters must be a list".to_string()))?;
+
+    /* WarningsState.filters could change while we are iterating over it. */
+    for i in 0..filters.borrow_vec().len() {
+        let tmp_item = filters.borrow_vec().get(i).cloned();
+        let tmp_item = if let Some(tmp_item) = tmp_item {
+            let tmp_item = PyTupleRef::try_from_object(vm, tmp_item)?;
+            if tmp_item.len() != 5 {
+                Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+            } else {
+                Ok(tmp_item)
+            }
+        } else {
+            Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+        }?;
+
+        /* Python code: action, msg, cat, mod, ln = item */
+        let _action = tmp_item.get(0);
+        let _msg = tmp_item.get(1);
+        let _cat = tmp_item.get(2);
+        let _item_mod = tmp_item.get(3);
+        let _ln_obj = tmp_item.get(4);
+    }
+
+    Ok(())
+}
+
+fn already_warned(
+    registry: PyObjectRef,
+    key: PyObjectRef,
+    should_set: usize,
+    vm: &VirtualMachine,
+) -> bool {
+    let version_obj = registry.get_item("version", vm).ok();
+    let filters_version = vm
+        .ctx
+        .new_int(vm.state.warnings.filters_version)
+        .as_object()
+        .to_owned();
+
+    if version_obj.is_none()
+        || version_obj.as_ref().unwrap().try_int(vm).is_err()
+        || !version_obj.unwrap().is(&filters_version)
+    {
+        let registry = registry.dict();
+        registry.as_ref().map(|registry| {
+            registry.into_iter().collect_vec().clear();
+            registry
+        });
+
+        if let Some(registry) = registry {
+            if registry.set_item("version", filters_version, vm).is_err() {
+                return false;
+            }
+        }
+    } else {
+        let already_warned = registry.get_item(key.as_ref(), vm);
+        return already_warned.is_ok();
```

Comment:
```suggestion
        return already_warned.is_ok();
```
```suggestion
        let already_warned = registry.get_item(key.as_ref(), vm)?;
        if already_warned.is_true(vm)? {
            return Ok(true);
        }
```
We need to return when the `already_warned` is true not for `Ok` in `PyResult`